rasa/core/actions/two_stage_fallback.py

```diff
@@ -0,0 +1,197 @@
+from rasa.core.trackers import DialogueStateTracker
+from rasa.utils.endpoints import EndpointConfig
+
+ACTION_TWO_STAGE_FALLBACK_NAME = "two_stage_fallback"
+
+
+class TwoStageFallbackAction(LoopAction):
+    def __init__(self, action_endpoint: Optional[EndpointConfig] = None) -> None:
+        self._action_endpoint = action_endpoint
+
+    def name(self) -> Text:
+        return ACTION_TWO_STAGE_FALLBACK_NAME
+
+    async def do(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> List[Event]:
+        if _user_should_affirm(tracker, events_so_far):
+            return await self._ask_affirm(output_channel, nlg, tracker, domain)
+
+        return await self._ask_rephrase(output_channel, nlg, tracker, domain)
+
+    async def _ask_affirm(
+        self,
+        output_channel: OutputChannel,
+        nlg: NaturalLanguageGenerator,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+    ) -> List[Event]:
+        affirm_action = action.action_from_name(
+            ACTION_DEFAULT_ASK_AFFIRMATION_NAME,
+            self._action_endpoint,
+            domain.user_actions,
+        )
+
+        return await affirm_action.run(output_channel, nlg, tracker, domain)
+
+    async def _ask_rephrase(
+        self,
+        output_channel: OutputChannel,
+        nlg: NaturalLanguageGenerator,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+    ) -> List[Event]:
+        rephrase = action.action_from_name(
+            ACTION_DEFAULT_ASK_REPHRASE_NAME, self._action_endpoint, domain.user_actions
+        )
+
+        return await rephrase.run(output_channel, nlg, tracker, domain)
+
+    async def is_done(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> bool:
+        _user_clarified = _last_intent_name(tracker) not in [
+            DEFAULT_NLU_FALLBACK_INTENT_NAME,
+            USER_INTENT_OUT_OF_SCOPE,
+        ]
+        return (
+            _user_clarified
+            or _two_fallbacks_in_a_row(tracker)
+            or _second_affirmation_failed(tracker)
+        )
+
+    async def deactivate(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> List[Event]:
+        if _two_fallbacks_in_a_row(tracker) or _second_affirmation_failed(tracker):
+            return await self._give_up(output_channel, nlg, tracker, domain)
+
+        return await self._revert_fallback_events(
+            output_channel, nlg, tracker, domain, events_so_far
+        ) + _message_clarification(tracker)
+
+    async def _revert_fallback_events(
```

Comment:
I think it is ok for now, but we'll need to rethink this revert strategy

Reply:
What are your concerns?

Reply:
because it is related to unfeaturization, and reverting is not the most elegant approach to solve it